pyserini/dsearch/__main__.py

```diff
@@ -0,0 +1,89 @@
+#
+# Pyserini: Python interface to the Anserini IR toolkit built on Lucene
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import argparse
+import os
+
+import numpy as np
+from tqdm import tqdm
+
+from pyserini.dsearch import QueryEncoder, SimpleDenseSearcher
+from pyserini.search import get_topics
+
+parser = argparse.ArgumentParser(description='Search a Faiss index.')
+parser.add_argument('--index', type=str, metavar='path to index or index name', required=True,
+                    help="Path to Faiss index or name of prebuilt index.")
+parser.add_argument('--topics', type=str, metavar='topic_name', required=True,
+                    help="Name of topics. Available: msmarco_passage_dev_subset.")
+parser.add_argument('--hits', type=int, metavar='num', required=False, default=1000, help="Number of hits.")
+parser.add_argument('--batch', type=int, metavar='num', required=False, default=1,
+                    help="search batch of queries in parallel")
+parser.add_argument('--msmarco',  action='store_true', default=False, help="Output in MS MARCO format.")
+parser.add_argument('--output', type=str, metavar='path', required=True, help="Path to output file.")
+args = parser.parse_args()
+
+topics = get_topics(args.topics)
+
+if os.path.exists(args.index):
+    # create searcher from index directory
+    searcher = SimpleDenseSearcher(args.index)
+else:
+    # create searcher from prebuilt index name
+    searcher = SimpleDenseSearcher.from_prebuilt_index(args.index)
+query_encoder = QueryEncoder(searcher.index_dir)
+
+if not searcher:
+    exit()
+
+# invalid topics name
+if topics == {}:
+    print(f'Topic {args.topics} Not Found')
+    exit()
+
+# build output path
+output_path = args.output
+
+print(f'Running {args.topics} topics, saving to {output_path}...')
+tag = 'Faiss'
+
+if args.batch > 1:
+    with open(output_path, 'w') as target_file:
+        topic_keys = sorted(topics.keys())
+        for i in tqdm(range(0, len(topic_keys), args.batch)):
+            topic_key_batch = topic_keys[i: i+args.batch]
+            topic_emb_batch = np.array([query_encoder.encode(topics[topic].get('title').strip())
```

Comment:
I got lost here reading the code... the query encodings are pre-stored somewhere right? Where are they loaded?

Should we also have something like `.load_encoded_queries()` or something like that to load pre-encoded queries?

Reply:
There is a line `query_encoder = QueryEncoder(searcher.index_dir)` where I loaded the pre encoded query from the `index dir` 
I place the pre encoded query with the pre build index rn, i.e. the registered `msmarco-passage-tct_colbert.tar.gz` contains `index`+`docid`+`query text`+`query embedding`.

will we always have a pre_encoded query for a prebuild index? If so, I feel we can pack prebuild index and pre encoded queries together like like what i am doing rn?

Reply:
I see. I think the pre-encoded queries should be separate, because an index can have multiple queries - for example, for MS MARCO, there's dev queries and test queries.

I think something like `.load_encoded_queries()` would be clearer.

Reply:
sure, sg